### PR TITLE
feat: bank reconciliation and plaid changes

### PR DESCRIPTION
**Pull request by casesolved-co-uk**
Many changes and bug fixes to bank reconciliation and plaid import to improve usability

- Proper reconciliation of a bank transaction against any number of accounting entries, including other bank transactions, for example when too much is transferred and there is a refund
- Vouchers are only cleared when fully reconciled
- Fully unreconcile a Bank Transaction
- Dialog usability including a fix to keep the list of bank transactions updated (bug)
- Proper plaid link refresh

> fix: plaid link refresh: update account ids
> fix: plaid transactions for credit cards & add accounts on link refresh if they don't exist
> fix: bank reconciliation amount matching
> fix: bank reconciliation dialog usability
> feat: rewrite bank transaction reconciliation to allow multiple transactions to reconcile against vouchers before clearance
> fix: matching transaction amounts and race condition bug
> fix: ensure there is a reference number in plaid transactions and other tweaks
> feat: add references to Payroll Entry Bank Journal Entry
> feat: only clear Voucher once all Bank GLEs are allocated to Bank Transactions
> fix: strange type error
> feat: add payment method field to bank and plaid transactions and prepopulate relevant bank reconciliation new voucher fields
> feat: bank reconciliation - allow bank transactions to reconcile against themselves for when there are banking amendments
> fix: bank transaction self-reconcile bug and tidy
> fix: bank reconciliation datatable index update

version-14-hotfix
version-13-hotfix

`no-docs`

### CODE REVIEW

**Comment by parsimonyERP** [comment]
Looks like wonderful work Richard!

File: erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.js [hunk 1]
```diff
@@ -155,7 +155,7 @@ frappe.ui.form.on("Bank Reconciliation Tool", {
 		}
 	},
 
-	render_chart: frappe.utils.debounce((frm) => {
+	render_chart(frm) {
```
Comment (barredterra) [inline erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.js:158]: Why was this changed? Looks like the original intention was to render the chart at most 2 times per second.

File: erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.py [hunk 2]
```diff
@@ -131,10 +131,8 @@ def create_journal_entry_bts(
 		fieldname=["name", "deposit", "withdrawal", "bank_account"],
 		as_dict=True,
 	)[0]
-	company_account = frappe.get_cached_value(
-		"Bank Account", bank_transaction.bank_account, "account"
-	)
-	account_type = frappe.get_cached_value("Account", second_account, "account_type")
+	company_account = frappe.get_value("Bank Account", bank_transaction.bank_account, "account")
```
Comment (barredterra) [inline erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.py:134]:
```suggestion
	company_account = frappe.db.get_value("Bank Account", bank_transaction.bank_account, "account")
```

For consistency

File: erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.py [hunk 3]
```diff
@@ -160,14 +156,12 @@ def create_journal_entry_bts(
...
-			if bank_transaction.withdrawal > 0
-			else 0,
-			"debit_in_account_currency": bank_transaction.deposit if bank_transaction.deposit > 0 else 0,
+			"credit_in_account_currency": bank_transaction.withdrawal,
+			"debit_in_account_currency": bank_transaction.deposit,
 		}
 	)
 
-	company = frappe.get_cached_value("Account", company_account, "company")
+	company = frappe.get_value("Account", company_account, "company")
```
Comment (barredterra) [inline erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.py:164]:
```suggestion
	company = frappe.db.get_value("Account", company_account, "company")
```

For consistency

File: erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.py [hunk 4]
```diff
@@ -220,12 +220,10 @@ def create_payment_entry_bts(
...
 	paid_amount = bank_transaction.unallocated_amount
-	payment_type = "Receive" if bank_transaction.deposit > 0 else "Pay"
+	payment_type = "Receive" if bank_transaction.deposit > 0.0 else "Pay"
 
-	company_account = frappe.get_cached_value(
-		"Bank Account", bank_transaction.bank_account, "account"
-	)
-	company = frappe.get_cached_value("Account", company_account, "company")
+	company_account = frappe.get_value("Bank Account", bank_transaction.bank_account, "account")
+	company = frappe.get_value("Account", company_account, "company")
```
Comment (barredterra) [inline erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.py:226]:
```suggestion
	company_account = frappe.db.get_value("Bank Account", bank_transaction.bank_account, "account")
	company = frappe.db.get_value("Account", company_account, "company")
```

For consistency

File: erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.py [hunk 5]
```diff
@@ -540,46 +516,82 @@ def get_matching_queries(
...
+			date AS posting_date,
+			currency
+		FROM
+			`tabBank Transaction`
+		WHERE
+			status != 'Reconciled'
+			AND name != '{transaction.name}'
+			AND bank_account = '{transaction.bank_account}'
+			AND {field} {'= %(amount)s' if exact_match else '> 0.0'}
+	"""
```
Comment (barredterra) [inline erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.py:591]: SQL should not be constructed in this way, because it is prone to SQL injection attacks. Maybe we can use a normal `frappe.get_all( ... )` and calculate the rank using python? Or rewrite it, using the query builder?

File: erpnext/accounts/doctype/bank_transaction/bank_transaction.js [hunk 6]
```diff
@@ -12,8 +12,13 @@ frappe.ui.form.on("Bank Transaction", {
...
 	},
-
-	bank_account: function(frm) {
+	refresh(frm) {
+		frm.add_custom_button('Unreconcile Transaction', () => {
+			frm.call('remove_payment_entries')
+			.then( () => frm.refresh() );
+		});
+	},
+	bank_account: function (frm) {
```
Comment (barredterra) [inline erpnext/accounts/doctype/bank_transaction/bank_transaction.js]: Format with _prettier_ for consistent style

**Review by barredterra — COMMENTED** [review]
Just a quick code review, didn't test the functionality yet. ✌🏼

File: erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.js (hunk 1 above)
Comment (casesolved-co-uk) [reply erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.js:158]: This is a hack because render_chart is called twice on load. That is a bug that needs to be fixed rather than a hack to hide the problem.
It's not an animation.

File: erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.py (hunk 5 above)
Comment (casesolved-co-uk) [reply erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.py:591]: Most of the other queries are written in SQL, and this was written before qb existed. It's tried and tested and I'm not spending the time rewriting it when the majority of changes are rejected anyway.

File: erpnext/accounts/doctype/bank_transaction/bank_transaction.js (hunk 6 above)
Comment (casesolved-co-uk) [reply erpnext/accounts/doctype/bank_transaction/bank_transaction.js]: If it fails the linting process, I'll reformat then.

**Comment by casesolved-co-uk** [comment]
@deepeshgarg007 now passing all tests 😁

File: erpnext/accounts/doctype/bank_transaction/bank_transaction.py [hunk 7]
```diff
@@ -18,72 +15,137 @@ def on_submit(self):
...
+	# nosemgrep: frappe-semgrep-rules.rules.frappe-modifying-but-not-comitting
 	def on_update_after_submit(self):
-		self.update_allocations()
-		self.clear_linked_payment_entries()
-		self.set_status(update=True)
+		"Run on save(). Avoid recursion caused by multiple saves"
+		if not self._saving_flag:
+			self._saving_flag = True
+			self.clear_linked_payment_entries()
+			self.update_allocations()
```
Comment (barredterra) [inline erpnext/accounts/doctype/bank_transaction/bank_transaction.py:26]: @deepeshgarg007 can we agree to not merge horrible things like this? 🥴
